src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1106,83 +1104,151 @@ TargetEntryChangesValue(TargetEntry *targetEntry, Var *column, FromExpr *joinTre
 {
-	Task *modifyTask = NULL;
-	DistTableCacheEntry *cacheEntry = DistributedTableCacheEntry(distributedTableId);
-
-	modifyTask = CitusMakeNode(Task);
-	modifyTask->jobId = INVALID_JOB_ID;
-	modifyTask->taskId = INVALID_TASK_ID;
-	modifyTask->taskType = MODIFY_TASK;
-	modifyTask->queryString = NULL;
-	modifyTask->anchorShardId = INVALID_SHARD_ID;
-	modifyTask->dependedTaskList = NIL;
-	modifyTask->replicationModel = cacheEntry->replicationModel;
+	Oid distributedTableId = ExtractFirstDistributedTableId(query);
+	List *taskList = NIL;
+	Job *job = NULL;
+	bool requiresMasterEvaluation = false;
+	bool deferredPruning = false;
 
-	if (originalQuery->onConflict != NULL)
+	if (!CanShardPrune(distributedTableId, query))
 	{
-		RangeTblEntry *rangeTableEntry = NULL;
+		/* there is a non-constant in the partition column, cannot prune yet */
+		taskList = NIL;
+		deferredPruning = true;
 
-		/* set the flag */
-		modifyTask->upsertQuery = true;
-
-		/* setting an alias simplifies deparsing of UPSERTs */
-		rangeTableEntry = linitial(originalQuery->rtable);
-		if (rangeTableEntry->alias == NULL)
+		/* must evaluate the non-constant in the partition column */
+		requiresMasterEvaluation = true;
+	}
+	else
+	{
+		taskList = RouterModifyTaskList(query, planningError);
+		if (*planningError)
 		{
-			Alias *alias = makeAlias(CITUS_TABLE_ALIAS, NIL);
-			rangeTableEntry->alias = alias;
+			return NULL;
 		}
+
+		/* determine whether there are function calls to evaluate */
+		requiresMasterEvaluation = RequiresMasterEvaluation(originalQuery);
 	}
 
-	if (shardInterval != NULL)
+	if (!requiresMasterEvaluation)
 	{
-		uint64 shardId = shardInterval->shardId;
-		StringInfo queryString = makeStringInfo();
+		/* no functions or parameters, build the query strings upfront */
+		RebuildQueryStrings(originalQuery, taskList);
+	}
+
+	job = CreateJob(originalQuery);
+	job->taskList = taskList;
+	job->requiresMasterEvaluation = requiresMasterEvaluation;
+	job->deferredPruning = deferredPruning;
+
+	return job;
+}
+
+
+/*
+ * CreateJob returns a new Job for the given query.
+ */
+static Job *
+CreateJob(Query *query)
+{
+	Job *job = NULL;
+
+	job = CitusMakeNode(Job);
+	job->dependedJobList = NIL;
+	job->dependedJobList = NIL;
+	job->jobId = INVALID_JOB_ID;
+	job->subqueryPushdown = false;
+	job->jobQuery = query;
+	job->taskList = NIL;
+	job->requiresMasterEvaluation = false;
+
+	return job;
+}
+
+
+/*
+ * CanShardPrune determines whether a query is ready for shard pruning
+ * by checking whether there is a constant value in the partition column.
+ */
+static bool
+CanShardPrune(Oid distributedTableId, Query *query)
+{
+	uint32 rangeTableId = 1;
+	Var *partitionColumn = NULL;
+	Expr *partitionValueExpr = NULL;
 
-		/* grab shared metadata lock to stop concurrent placement additions */
-		LockShardDistributionMetadata(shardId, ShareLock);
+	if (query->commandType != CMD_INSERT)
+	{
+		/* we assume UPDATE/DELETE is always prunable */
```

Comment:
In short (I think), applying a `DELETE` or `UPDATE` on a larger set of shards than is strictly necessary should not have any ill effects (overhead, locking, etc. aside). The pruning is an optimization.

But if you apply an `INSERT` to too many shards, you're in for trouble.

So I'm guessing the comment means mainly "even if we can't prune, we can just return all shards". This would error out now presumably anyhow (though as @marcocitus pointed out, that need not be true in the near future).